core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1490,6 +1495,39 @@ class Log(@volatile var dir: File,
     }
   }
 
+  private def recordConsumerFetchLagMetrics(segmentBaseOffset: Long, segmentSizeInBytes: Long, 
+                                    endOffset: Long, fetchInfo: FetchDataInfo){   
+    // record fetch time lag in milliseconds 
+    // time lag = activeSegment.maxTimestampSoFar - MaximumTimestamp of first batch fetched for the startOffset  
+    val iterator = fetchInfo.records.batches.iterator
+    if (iterator.hasNext()){
+      val batch = iterator.next()
+      if (batch.magic >= RecordBatch.MAGIC_VALUE_V2) {
+        val startOffsetTimestamp = batch.maxTimestamp()
+        val maxTimestampSoFar = activeSegment.maxTimestampSoFar
+        if (startOffsetTimestamp != -1 && maxTimestampSoFar != -1) {
+          val fetchLagInMS = maxTimestampSoFar - startOffsetTimestamp
+          brokerTopicStats.allTopicsStats.consumerFetchLagTimeInMs.update(fetchLagInMS)
+        }
+      }
+    } 
+    
+    // record fetch lag in bytes
+    // consumerFetchLagInBytes = Sum of the newer segments (if any) + 
+    //                          (segment_size - relative position of fetched offset in segment)
+    val consumerFetchLagInBytes = {
+      val nextSegmentEntry = segments.higherEntry(segmentBaseOffset)
+      val segmentLag = segmentSizeInBytes - fetchInfo.fetchOffsetMetadata.relativePositionInSegment.toLong
+      if (nextSegmentEntry == null) {
+         segmentLag
+      }else {
+         val newerSegmentsSize = logSegments(nextSegmentEntry.getKey, endOffset).map(_.size.toLong).sum
```

Comment:
This code seems fine, as long as an additional logSegments(from, to) in the fetch path is not significant. I have seen the ConcurrentNavigableMap code turn up in profiling, so that's my main concern.

Reply:
Yes, the reason of calculating the the tail end is because the size calculation is O(N), where N is the number of lagging (/newer) segments .

Reply:
We should still test the bad case. As @ijuma said above, I think we need a jmh benchmark for this code.

Reply:
totally agree, we will have benchmark testing as a requirement, I will add results once ready